src/downloader/headers/verify_stage.rs

```diff
@@ -63,8 +72,62 @@ impl VerifyStage {
         })
     }
 
-    fn verify_slice(&self, _slice: &HeaderSlice) -> bool {
-        // TODO: verify hashes properly
-        rand::random::<u8>() < 224
+    /// The algorithm verifies that the top of the slice matches one of the preverified hashes,
+    /// and that all blocks down to the root of the slice are connected by the parent_hash field.
+    ///
+    /// For example, if we have a HeaderSlice[192...384]
+    /// (with block headers from 192 to 384 inclusive), it verifies that:
+    ///
+    /// hash(slice[384]) == preverified hash(384)
+    /// hash(slice[383]) == slice[384].parent_hash
+    /// hash(slice[382]) == slice[383].parent_hash
+    /// ...
+    /// hash(slice[192]) == slice[193].parent_hash
+    ///
+    /// Thus verifying hashes of all the headers.
+    fn verify_slice(&self, slice: &HeaderSlice) -> bool {
+        if slice.headers.is_none() {
+            return false;
+        }
+        let headers = slice.headers.as_ref().unwrap();
```

Comment:
can this actually panic? If so, we might want to offer a more comprehensive error message.

Reply:
It can't, I'm checking for None 2 lines above. I prefer to not use match to not increase nestedness level more than necessary (some languages have a `guard` construct for this, which Rust doesn't have)

Reply:
you can use `expect("some msg")`

Reply:
`.expect("not null, see the line above")` ?
:trollface: 😄

Reply:
I wish Rust had something [like this](https://docs.swift.org/swift-book/ReferenceManual/Statements.html#ID524)

Reply:
see also: https://softwareengineering.stackexchange.com/questions/18454/should-i-return-from-a-function-early-or-use-an-if-statement

Reply:
.expect("not None, see the line above") ?
:trollface: 😄

I wish Rust had something [like this (inverse of "if let")](https://docs.swift.org/swift-book/ReferenceManual/Statements.html#ID524)

see also: https://softwareengineering.stackexchange.com/questions/18454/should-i-return-from-a-function-early-or-use-an-if-statement

Reply:
haha, my point is:
if it cannot panic, then there is no need for an option. And if we're using `unwrap()` is because we _know_ it should not panic. If it does, then which invariant is not holding?

Reply:
@blasrodri there's a need for option, because it can be in 2 states "is_none" or "is_some", but since I already check "is_none", and since the slice is a ref here, it can't be mutated between the lines, so on the unwrap line it is not possible, and it is guaranteed to be `is_some()`.

😮 the Rust gods listened to us ✨ :
https://github.com/rust-lang/rfcs/pull/3137

it mentions nightly-2021-09-04, so might become available to us soon? 🤔 

with this compiler change in place this piece of code could be rewritten to:

    let Some(headers) = slice.headers.as_ref() else { return }

Reply:
It has become available to us now 😉

Reply:
I've tried the new syntax like so:

```
diff --git a/src/downloader/headers/verify_stage.rs b/src/downloader/headers/verify_stage.rs
index f760307..68bd69e 100644
--- a/src/downloader/headers/verify_stage.rs
+++ b/src/downloader/headers/verify_stage.rs
@@ -85,17 +85,11 @@ impl VerifyStage {
     /// hash(slice[192]) == slice[193].parent_hash
     ///
     /// Thus verifying hashes of all the headers.
+    #[allow(clippy::needless_return)]
     fn verify_slice(&self, slice: &HeaderSlice) -> bool {
-        if slice.headers.is_none() {
-            return false;
-        }
-        let headers = slice.headers.as_ref().unwrap();
-
-        if headers.is_empty() {
-            return true;
-        }
+        let Some(headers) = slice.headers.as_ref() else { return false; };
+        let Some(last) = headers.last() else { return true; };
 
-        let last = headers.last().unwrap();
         let last_hash = last.hash();
         let expected_last_hash =
             self.preverified_hash(slice.start_block_num + headers.len() as u64 - 1);
diff --git a/src/lib.rs b/src/lib.rs
index 707b5e5..e3cc647 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -3,6 +3,7 @@
     destructuring_assignment,
     entry_insert,
     generic_associated_types,
+    let_else,
     linked_list_cursors,
     never_type,
     type_alias_impl_trait
```

But found 2 problems:

1. My IDE does not support it yet, so the code in this method becomes erroneous:
![Screenshot 2021-09-12 at 11 28 55](https://user-images.githubusercontent.com/11477595/132982452-976805d8-6d0e-4b25-b8f3-3d97d3ba1ba9.png)
This will be fixed eventually: https://github.com/intellij-rust/intellij-rust/issues/7773

2. Clippy reports needless_return by mistake. If I remove the return, it fails to compile.
![Screenshot 2021-09-12 at 11 29 47](https://user-images.githubusercontent.com/11477595/132982481-e1c7f7fc-b50f-49bb-984b-ccd4dcd1138f.png)

I'd rather keep this as is for now, and change to the new syntax when it becomes more widely supported.